nio/client/base_client.py

```diff
@@ -129,11 +130,27 @@ class ClientCallback:
     func: Union[Callable[..., None], Callable[..., Awaitable[None]]] = field()
     filter: Union[Tuple[Type, ...], Type, None] = None
 
-    async def execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+    def _execute(self, event, room: Optional[MatrixRoom] = None) -> Optional[Awaitable]:
+        # Checks the filter and executes the function once.
+        # sync_execute and async_execute will each determine
+        # how to run an awaitable if one is returned.
         if self.filter is None or isinstance(event, self.filter):
-            result = self.func(room, event) if room else self.func(event)
-            if inspect.isawaitable(result):
-                await result
+            if room:
+                return self.func(room, event)
+            else:
+                return self.func(event)
+
+    def sync_execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+        """Execute callback from synchronous context."""
+        result = self._execute(event, room)
+        if inspect.isawaitable(result):
+            asyncio.run(result)
```

Comment:
It seems like that the main motivation to have a separate `sync_execute` and `async_execute` is right here: the ability to support async callbacks in a sync client.

Considering the extra complexity it introduces, and considering that it involves implicitly supporting/adding new features to a soft-deprecated usecase (using `matrix-nio` without async) I'm tempted to say that we might not care to include this extra complexity, and stick to the original implementation of `execute` here.

I don't generally expect people who know how to write `async def` callback functions to still be using the sync client, as well.

Simultaneously, I don't think this is a bad change and it's fairly straightforward. So I'm more tempted to keep it.